server/src/test/java/io/spine/examples/shareaware/server/e2e/SharePurchaseTest.java

```diff
@@ -0,0 +1,139 @@
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.chooseTeslaShareFrom;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.insufficientFundsAfter;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.investmentAfterTeslaPurchase;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.walletBalanceWith;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.zeroWalletBalance;
+import static java.time.Duration.ofMillis;
+
+/**
+ * End-to-end test that describes such a scenario:
+ * <ol>
+ *     <li>The user attempts to purchase shares with no money in the wallet.</li>
+ *     <li>The user replenishes his wallet for 500 dollars.</li>
+ *     <li>The user successfully purchases one 'Tesla' share.</li>
+ *     <li>The user withdraws all his money from the wallet.</li>
+ * </ol>
+ */
+final class SharePurchaseTest extends WithClient {
+
+    @Test
+    @DisplayName("User should purchase one tesla share and withdraw all the money after this")
+    void test() throws ExecutionException, InterruptedException {
+        SharePurchaseUser user = new SharePurchaseUser(client());
+
+        WalletBalance initialBalance = user.signsUp();
+        sleepUninterruptibly(ofMillis(1500));
+        List<Share> shares = user.looksAtShares();
+        Share tesla = chooseTeslaShareFrom(shares);
+
+        WalletBalance balanceAfterFailedPurchase = user.attemptsToPurchaseShare(tesla);
+        assertThat(balanceAfterFailedPurchase).isEqualTo(initialBalance);
+
+        WalletBalance balanceAfterReplenishment = user.replenishesWalletFor(usd(500));
+        WalletBalance expectedBalanceAfterReplenishment =
+                walletBalanceWith(usd(500), user.walletId());
+        assertThat(balanceAfterReplenishment).isEqualTo(expectedBalanceAfterReplenishment);
+
+        SubscriptionOutcome<WalletBalance> reducedBalance =
+                user.expectsChangesIn(WalletBalance.class);
+        InvestmentView investmentInTesla = user.purchasesShares(tesla, 1);
+        InvestmentView expectedInvestmentInTesla = investmentAfterTeslaPurchase(tesla, user.id());
+        WalletBalance balanceAfterPurchase = user.checksChangesIn(reducedBalance);
+        WalletBalance expectedBalanceAfterPurchase =
+                balanceAfterTeslaPurchase(tesla.getPrice(), balanceAfterReplenishment);
+        assertThat(investmentInTesla).isEqualTo(expectedInvestmentInTesla);
+        assertThat(balanceAfterPurchase).isEqualTo(expectedBalanceAfterPurchase);
+
+        WalletBalance walletAfterWithdrawal = user.withdrawsAllHisMoney(balanceAfterPurchase);
+        assertThat(walletAfterWithdrawal).isEqualTo(zeroWalletBalance(user.walletId()));
+    }
+
+    /**
+     * The user for a {@link SharePurchaseTest} that can perform actions
+     * that describe the test scenario.
+     */
+    private class SharePurchaseUser extends E2EUser {
```

Comment:
Let's pull this type up to the package level.

The code will be cleaner this way.